cmd/juju/cloud/add_test.go

```diff
@@ -7,172 +7,256 @@ import (
+//       new-york:
+//         endpoint: http://newyork/1.0
+//   garage-maas:
+//     type: mass
+//     auth-types: [oauth]
+//     endpoint: http://garagemaas
+// `[1:]
+// 	err = ioutil.WriteFile(sourceFile, []byte(source), 0600)
+// 	c.Assert(err, jc.ErrorIsNil)
+// 	return sourceFile
+// }
+
 func (s *addSuite) TestAddBadFilename(c *gc.C) {
-	addCmd := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	badFileErr := errors.New("")
+	fake.Call("ParseCloudMetadataFile", "somefile.yaml").Returns(map[string]cloudfile.Cloud{}, badFileErr)
+
+	addCmd := cloud.NewAddCloudCommand(fake)
 	_, err := testing.RunCommand(c, addCmd, "cloud", "somefile.yaml")
-	c.Assert(err, gc.ErrorMatches, "open somefile.yaml: .*")
+	c.Check(err, gc.Equals, badFileErr)
 }
 
 func (s *addSuite) TestAddBadCloudName(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	addCmd := cloud.NewAddCloudCommand()
-	_, err := testing.RunCommand(c, addCmd, "cloud", sourceFile)
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "testFile").Returns(map[string]cloudfile.Cloud{}, nil)
+
+	addCmd := cloud.NewAddCloudCommand(fake)
+	_, err := testing.RunCommand(c, addCmd, "cloud", "testFile")
 	c.Assert(err, gc.ErrorMatches, `cloud "cloud" not found in file .*`)
 }
 
 func (s *addSuite) TestAddExisting(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(), "homestack", sourceFile)
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "fake.yaml").Returns(homestackMetadata(), nil)
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+
+	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(fake), "homestack", "fake.yaml")
 	c.Assert(err, gc.ErrorMatches, `"homestack" already exists; use --replace to replace this existing cloud`)
 }
 
 func (s *addSuite) TestAddExistingReplace(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(), "homestack", sourceFile, "--replace")
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "fake.yaml").Returns(homestackMetadata(), nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+	numCallsToWrite := fake.Call("WritePersonalCloudMetadata", homestackMetadata()).Returns(nil)
+
+	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(fake), "homestack", "fake.yaml", "--replace")
 	c.Assert(err, jc.ErrorIsNil)
-	data, err := ioutil.ReadFile(osenv.JujuXDGDataHomePath("clouds.yaml"))
-	c.Assert(string(data), gc.Equals, `
-clouds:
-  homestack:
-    type: openstack
-    auth-types: [userpass, access-key]
-    endpoint: http://homestack
-    regions:
-      london:
-        endpoint: http://london/1.0
-      new-york:
-        endpoint: http://newyork/1.0
-`[1:])
+
+	c.Check(numCallsToWrite(), gc.Equals, 1)
```

Comment:
If we write out the wrong value, the call will never be made and this function will return 0. You're correct that the arguments are tested implicitly. I think I disagree that this is a bad thing?